waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -50,52 +51,71 @@ validateExactlyOneAppExists spec =
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'username' of type 'string'."
-                    ],
-                case maybePasswordField of
-                  Just passwordField
-                    | Entity.Field.fieldType passwordField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'password' of type 'string'."
-                    ]
+         in concatMap
+              (validateField "app.auth.userEntity" userEntityFields)
+              [ ("username", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                ("password", Entity.Field.FieldTypeScalar Entity.Field.String, "String")
               ]
 
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed :: AppSpec -> [ValidationError]
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed spec = case App.auth (snd $ getApp spec) of
+  Nothing -> []
+  Just auth ->
+    if not $ Auth.isExternalAuthEnabled auth
+      then []
+      else case Auth.externalAuthAssociationEntity auth of
+        Nothing -> [GenericValidationError "Expected an Entity referenced by app.auth.externalAuthAssociationEntity to exist when using a social login method."]
+        Just externalAuthAssociationEntityRef ->
+          let (userEntityName, userEntity) = AS.resolveRef spec (Auth.userEntity auth)
+              userEntityFields = Entity.getFields userEntity
+              (externalAuthAssociationEntityName, externalAuthAssociationEntity) = AS.resolveRef spec externalAuthAssociationEntityRef
+              externalAuthAssociationEntityFields = Entity.getFields externalAuthAssociationEntity
+           in concatMap
+                (validateField "app.auth.externalAuthAssociationEntity" externalAuthAssociationEntityFields)
+                [ ("provider", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                  ("providerId", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                  ("user", Entity.Field.FieldTypeScalar (Entity.Field.UserType userEntityName), userEntityName),
+                  ("userId", Entity.Field.FieldTypeScalar Entity.Field.Int, "Int")
+                ]
+                ++ concatMap
+                  (validateField "app.auth.userEntity" userEntityFields)
+                  [ ("externalAuthAssociations", Entity.Field.FieldTypeComposite $ Entity.Field.List $ Entity.Field.UserType externalAuthAssociationEntityName, externalAuthAssociationEntityName ++ "[]")
+                  ]
+
+validateField :: String -> [Entity.Field.Field] -> (String, Entity.Field.FieldType, String) -> [ValidationError]
+validateField entityName entityFields (fieldName, fieldType, fieldTypeName) =
```

Comment:
So it checks that entity has a field of given name and type.
Maybe we could mention that in the comment?
Or maybe rename it to `validateEntityHasField` or `checkEntityHasField`?